### DISCUSSION TITLE
update state with data received from websocket

### DISCUSSION THREAD

**Discussion by iot49** [lit/lit#4424] [Q&A]
My app receives state updates from a websocket. I am trying to use lit context to make this state available to custom components.

Question: how do I update the context provider (declared in the root element) with new data?

Presently I am using an ugly workaround which supplies a pointer to the root element in the constructor of the context which the latter calls whenever updates are received. Is there a "cleaner" way to do this?

**Comment by augustjk** [comment]
I'm not sure what you mean by "supplies a pointer to the root element in the constructor of the context".

Updating the provided value can be done by setting the `@provide()` decorated property, or calling `setValue()` on the `ContextProvider` controller instance.

So maybe something like..
```ts
@customElement('my-provider)
class MyProvider extends LitElement {
  @provide({context})
  data;

  // or with a controller
  provider = new ContextProvider({context});

  socket: WebSocket;

  connectedCallback() {
    this.socket = new WebSocket('...');

    this.socket.addEventListener('message', (event) => {
      this.data = event.data;
      // or with a controller
      this.provider.setValue(event.data);
    });
  }

  disconnectedCallback() {
    this.socket.close();
  }

}
```

**Comment by iot49** [comment]
Thank you, `setValue` appears to be the solution.

But now I run into a new **problem**: consumers never get the updates from `setValue`. 

I assume the provider and consumers are "linked" via the context variable, declared in a shared file that and imported by both (following the documentation):

```
export const testContext = createContext<ITest>(Symbol('test'));
```

This works with decorators (but then I do not know what to call `setValue` on). Conversely, my consumers are not updated if switch to a manually created provider (`new ContextProvider(this, ...)`). 

Is there a working example, to either

- call `setValue` on a provider created with`@provide` or
- get consumers updated by a provider created with `new ContextProvider(this, ...)`

Many thanks!

**Comment by iot49** [comment]
For reference:

Found this: https://github.com/lit/lit/discussions/3685. Way over my head, but if I understand this correctly, it's difficult use setValue in lit, presently.

**Comment by iot49** [comment]
Got it working:

1. use `provider.setValue(..., true)` to force an update
2. `@consume({ context: ..., subscribe: true })` - I'd forgotten to set `subscribe` to true.
